posts/frangeris/03.secure-values-by-private-keys-in-aws.md

```diff
@@ -0,0 +1,235 @@
+
+![Run npx serverless command to start](./assets/03.image.1.png)
+
+Some operations used on the lambda function will require some permissions later, due to this, it's recommended to include and enable those permissions before the IAM role of the lambda gets created, here is what permissions we need the lambda to have:
+
+```yaml
+provider:
+  # ...
+  iam:
+    role:
+      statements:
+        - Effect: Allow
+          Action:
+            - "s3:*"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "cloudwatch:*"
+            - "logs:*"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "ssm:GetParameter"
+            - "ssm:GetParametersByPath"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "secretsmanager:GetSecretValue"
+          Resource: "*"
+        - Effect: Allow
+          Action:
+            - "kms:Decrypt"
+          Resource: "*"
+```
+
+Then, with your [aws account](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) already set up, let's deploy it.
+
+```bash
+npx sls deploy
+```
+
+🎉 After a correct deployment, the code should be created successfully and ready to use, now we need to configure the most important part, security.
+
+## Security
+
+Permission will be managed via [IAM policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) with the objective of _only allows the lambda to get and decrypt the secured value_, the approach used will be based on: **prevent IAM entities from accessing the KMS key and allow the root user account to manage it** (also preventing the root user account from losing access to the KMS key).
+
+This "protection" layer of our sensitive value is based on the composition of two AWS services, [Key Management Service](https://aws.amazon.com/kms/) is the one responsible for creating the private key which will be used to encrypt our value, then [Secret Manager](https://aws.amazon.com/es/secrets-manager/) or [System Manager: Parameters Store](https://aws.amazon.com/systems-manager/), both services allows us to accomplish the same functionality but differ on some caveats, _the chosen one_ will be used for saving the secured encrypted value, here is a quick graphic representation of the implementation:
+
+![A diagram that explains how services are used](./assets/03.image.2.png)
+
+Now lets translate these concepts into code, the first thing we need after having the lambda functional and usable is to create the key used to encrypt our _precious value_...
+
+![Meme of our precious value](./assets/03.image.3.png)
+
+So lets create the key using the "root" account and give access to lambda just for `"kms:Decrypt"`, not other operations!
```

Comment:
```suggestion
So lets create the key using the "root" account and give it access to lambda just for `"kms:Decrypt"`, not other operations!
```

Reply:
This is debatable. I think the lambda is the one getting the access, am I right?

Reply:
You're right, thanks!